src/python/pants/backend/scala/dependency_inference/scala_parser.py

```diff
@@ -115,6 +115,7 @@ class ScalaSourceDependencyAnalysis:
     provided_symbols_encoded: FrozenOrderedSet[str]
     imports_by_scope: FrozenDict[str, tuple[ScalaImport, ...]]
     consumed_symbols_by_scope: FrozenDict[str, FrozenOrderedSet[str]]
+    scopes: FrozenOrderedSet[str]
```

Comment:
Hm: we already have the relevant scopes via `imports_by_scope` and `consumed_symbols_by_scope`, don't we? In particular, to resolve #13655, it seems like we want to look up a symbol in any of its enclosing scopes, which actually requires the structure that we already have there. See the TODO I linked.

Reply:
We don't. The scope key in `consumed_symbols_by_scope` is the scope where the consumed symbol was used, _not_ the scopes that could potentially provide the symbol.

Moreover, take this example:

```
package foo
package bar.grok
```

If the file has no consumed symbols in `foo`, then `foo` won't show up as a scope key in `consumed_symbols_by_scope`.

Reply:
> We don't. The scope key in `consumed_symbols_by_scope` is the scope where the consumed symbol was used, _not_ the scopes that could potentially provide the symbol.

But the symbol could only possibly be provided by one of the parent scopes, right? Where a symbol is consumed is relevant.

For example, in:
```scala
package a {
  package b {
    class A {
      ...
    }
  }
  class C {
    ...
  }
}
```
... `class A` should not be visible to `class C`.

----

So is the idea just to walk upward from `consumed_symbols_by_scope` in the `scopes`, and suffix the symbol on each parent scope? In future, it might make things easier to include the consuming code in the same change, as it would answer these questions directly.

Reply:
> So is the idea just to walk upward from consumed_symbols_by_scope in the scopes, and suffix the symbol on each parent scope? 

No. In this example, let's say that `class A` references `X`. Then that symbol could be `a.b.X` or `a.X`.

Also, I'm not understanding why you mentioned "... class A should not be visible to class C." Am I missing something?

Reply:
> Also, I'm not understanding why you mentioned "... class A should not be visible to class C." Am I missing something?

Sorry, I meant: should not be in scope. i.e., `C` would not be able to refer to `A` without qualification. It would need to refer to it as absolute: `a.b.A`. So: visible, but not automatically in scope.

Reply:
> So is the idea just to walk upward from consumed_symbols_by_scope in the scopes, and suffix the symbol on each parent scope? 

Rethinking this, that is probably what I'll end up doing. Will handle that in the PR to actually update the dep inference logic.